### Control RNA-seq data

Before producing the paired RNA- and ATAC-seq dataset it was important to determine that the left and right somites are indeed equivalent and do not show any significant differences in their expression profiles.

To this end, we sequenced four pairs of somites from two different embryos from the same litter (and thus roughly from the same somite stage).

Additionally, for one pair of somites, we repeated the library prep procedure using only one quarter of the recommended reagent quantity. If the two libraries are equivalent we could then save on reagents by four-fold.

These libraries were sequenced in one lane of an Illumina HiSeq 4000, to produce single-end 50bp fragments.

#### Mapping the sequencing data

Since the libraries were sequenced on one lane only there is no need to merge FASTq files. We can directly proceed to map the sequencing reads to the mouse reference genome, following the same method as with the main dataset (see `01.0_mapping_RNAseq_reads.md` for details).

We map:

```{bash}
sh 02.1_align.sh
```

Compute mapping statistics:

```{bash}
perl 01.3_mappingStats.pl -dir /user01/group_folders/Personal/Ximena/somitogenesis2020/RNA-seq/data/STAR/CONTROL/ > /user01/group_folders/Personal/Ximena/somitogenesis2020/RNA-seq/data/mappingStatistics_CONTROLlibraries.tsv
```

And create a counts matrix:

```{bash}
R CMD BATCH 02.2_countMatrix.R
```
